### PR TITLE
Dimension Computation benchmark

### PR DESCRIPTION
**Pull request by brightchen**

### CODE REVIEW

**Comment by brightchen** [comment]
@vrozov Please review

File: apex-benchmarks/src/main/java/apex/benchmark/ProcessTimeAwareStore.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+import org.apache.apex.malhar.lib.dimensions.DimensionsEvent.Aggregate;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+import com.datatorrent.api.Context.OperatorContext;
+import com.datatorrent.contrib.dimensions.AppDataSingleSchemaDimensionStoreHDHT;
+
+
+public class ProcessTimeAwareStore extends AppDataSingleSchemaDimensionStoreHDHT
```
Comment (tweise) [inline apex-benchmarks/src/main/java/apex/benchmark/ProcessTimeAwareStore.java:29]: Please javadoc why this is needed for the benchmark.

**Review by tweise — CHANGES_REQUESTED** [review]

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java [hunk 2]
```diff
@@ -106,28 +112,52 @@ public void populateDimensionsDAG(DAG dag, Configuration conf, DefaultOutputPort
...
+    } else {
+      DevNull devNull = new DevNull();
+      dag.addOperator("devNull", devNull);
+      dag.addStream("QueryResult", store.queryResult, devNull.data);
+    }
+    
+    dag.setAttribute(DAGContext.STREAMING_WINDOW_SIZE_MILLIS, STREAMING_WINDOW_SIZE_MILLIS);
   }
   
+  protected static class StoreStreamCodec extends KryoSerializableStreamCodec<Aggregate>
```
Comment (vrozov) [inline apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java]: Why is KryoSerializableStreamCodec necessary here?

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java [hunk 3]
```diff
@@ -106,28 +112,52 @@ public void populateDimensionsDAG(DAG dag, Configuration conf, DefaultOutputPort
...
-    dag.setAttribute(dimensions, Context.OperatorContext.APPLICATION_WINDOW_COUNT, 10);
+    dag.setUnifierAttribute(dimensions.output, OperatorContext.MEMORY_MB, 10240);
+//    dag.setAttribute(dimensions, Context.OperatorContext.APPLICATION_WINDOW_COUNT, 120);
+//    dag.setAttribute(dimensions, Context.OperatorContext.CHECKPOINT_WINDOW_COUNT, 120);
+    
     dag.setInputPortAttribute(dimensions.input, Context.PortContext.PARTITION_PARALLEL, true);
     
     // store
     AppDataSingleSchemaDimensionStoreHDHT store = createStore(dag, conf, eventSchema); 
+    store.setCacheWindowDuration(10000 * 5 / STREAMING_WINDOW_SIZE_MILLIS);   //cache for 5 windows
```
Comment (vrozov) [inline apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java:128]: The number seems odd to me. Is the comment still accurate?

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java [hunk 4]
```diff
@@ -106,28 +112,52 @@ public void populateDimensionsDAG(DAG dag, Configuration conf, DefaultOutputPort
     dimensions.setConfigurationSchemaJSON(eventSchema);
     dimensions.setUnifier(new DimensionsComputationUnifierImpl<InputEvent, Aggregate>());
 
-    dag.setUnifierAttribute(dimensions.output, OperatorContext.MEMORY_MB, 4096);
-    dag.setAttribute(dimensions, Context.OperatorContext.APPLICATION_WINDOW_COUNT, 10);
+    dag.setUnifierAttribute(dimensions.output, OperatorContext.MEMORY_MB, 10240);
+//    dag.setAttribute(dimensions, Context.OperatorContext.APPLICATION_WINDOW_COUNT, 120);
```
Comment (vrozov) [inline apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java]: Please cleanup

File: apex-benchmarks/src/main/java/apex/benchmark/DimensionTuple.java [hunk 5]
```diff
@@ -18,10 +18,15 @@
...
   public long eventTime;
-  public long clicks;
+  public long clicks = 1;
+  
+  protected static boolean autoGeneratorCampaignId = true;
 
   public static DimensionTuple fromTuple(Tuple tuple)
   {
+    if(tuple.campaignId == null && autoGeneratorCampaignId) {
+      tuple.campaignId = "1111111111111111111";
```
Comment (vrozov) [inline apex-benchmarks/src/main/java/apex/benchmark/DimensionTuple.java:33]: Does not look like a valid campaignId.

File: apex-benchmarks/src/test/resources/log4j.properties [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+# to enable, add SYSLOG to rootLogger
+log4j.appender.SYSLOG=org.apache.log4j.net.SyslogAppender
+log4j.appender.SYSLOG.syslogHost=127.0.0.1
+log4j.appender.SYSLOG.layout=org.apache.log4j.PatternLayout
+log4j.appender.SYSLOG.layout.conversionPattern=${dt.cid} %-5p [%t] %c{2} %x - %m%n
+log4j.appender.SYSLOG.Facility=LOCAL1
+
+log4j.logger.org=info
+#log4j.logger.org.apache.commons.beanutils=warn
+log4j.logger.com.datatorrent=debug
```
Comment (vrozov) [inline apex-benchmarks/src/test/resources/log4j.properties:42]: Please review log level settings. DEBUG is good for development and testing. Benchmark should be done with ERROR or at least INFO.

**Review by vrozov — CHANGES_REQUESTED** [review]

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java (hunk 3 above)
Comment (brightchen) [reply apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java:128]: Our cache will cleaup all when timerout. So it's better to main large duration to decrease the chance of reload data of last window. But too large cache can impact the performance of locating. I choose this number based on some test result.

File: apex-benchmarks/src/main/java/apex/benchmark/DimensionTuple.java (hunk 5 above)
Comment (brightchen) [reply apex-benchmarks/src/main/java/apex/benchmark/DimensionTuple.java:33]: The campaignId should not empty in normal case. 
It used for unit test where campainId will not populated. choose a invalid campainId for easier identify if any problem

File: apex-benchmarks/src/test/resources/log4j.properties (hunk 6 above)
Comment (brightchen) [reply apex-benchmarks/src/test/resources/log4j.properties:42]: This is use for unit test

File: apex-benchmarks/src/main/java/apex/benchmark/DimensionTuple.java (hunk 5 above)
Comment (vrozov) [reply apex-benchmarks/src/main/java/apex/benchmark/DimensionTuple.java:33]: Please add comments explaining unit test usage

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java (hunk 3 above)
Comment (vrozov) [reply apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java:128]: It is not caching for 5 windows that will correspond clearing cache every second. The cache seems to be cleared every 13 hours.

Comment (brightchen) [reply apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java:128]: The expiry time is: 10000 \* 5 / STREAMING_WINDOW_SIZE_MILLIS \* STREAMING_WINDOW_SIZE_MILLIS  = 50000 mill-seconds = 5 application windows

**Comment by brightchen** [comment]
@vrozov please review

File: apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java (hunk 3 above)
Comment (brightchen) [reply apex-benchmarks/src/main/java/apex/benchmark/ApplicationDimensionComputation.java:128]: In the application, the application window in fact same as stream window. The previous application window meant the window of 10 second

**Pull request merged**